pkg/reconciler/kubequota/kubequota_controller.go

```diff
@@ -0,0 +1,366 @@
+}
+
+// NewController creates a new Controller.
+func NewController(
+	clusterWorkspacesInformer tenancyinformers.ClusterWorkspaceInformer,
+	kubeClusterClient kubernetes.ClusterInterface,
+	kubeInformerFactory informers.SharedInformerFactory,
+	dynamicDiscoverySharedInformerFactory *informer.DynamicDiscoverySharedInformerFactory,
+	crdInformer apiextensionsinformers.CustomResourceDefinitionInformer,
+	quotaRecalculationPeriod time.Duration,
+	fullResyncPeriod time.Duration,
+	workersPerLogicalCluster int,
+	informersStarted <-chan struct{},
+) (*Controller, error) {
+	c := &Controller{
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+
+		dynamicDiscoverySharedInformerFactory: dynamicDiscoverySharedInformerFactory,
+		kubeClusterClient:                     kubeClusterClient,
+		informersStarted:                      informersStarted,
+
+		quotaRecalculationPeriod: quotaRecalculationPeriod,
+		fullResyncPeriod:         fullResyncPeriod,
+
+		workersPerLogicalCluster: workersPerLogicalCluster,
+
+		cancelFuncs: map[logicalcluster.Name]func(){},
+
+		scopingGenericSharedInformerFactory: newScopingGenericSharedInformerFactory(dynamicDiscoverySharedInformerFactory),
+		scopingResourceQuotaInformer:        NewScopingResourceQuotaInformer(kubeInformerFactory.Core().V1().ResourceQuotas()),
+
+		crdsSynced: crdInformer.Informer().HasSynced,
+
+		getClusterWorkspace: func(key string) (*tenancyv1alpha1.ClusterWorkspace, error) {
+			return clusterWorkspacesInformer.Lister().Get(key)
+		},
+
+		listCRDs: func() ([]*apiextensionsv1.CustomResourceDefinition, error) {
+			return crdInformer.Lister().List(labels.Everything())
+		},
+	}
+
+	clusterWorkspacesInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: c.enqueue,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				c.enqueue(oldObj)
+			},
+			DeleteFunc: c.enqueue,
+		},
+	)
+
+	// This sets up a single global event handler on ResourceQuotas. When a resource quota controller is created for
+	// a logical cluster, it registers its own event handler, which is stored in Controller.resourceQuotaEventHandlers.
+	// When an event comes in, this handler looks up the actual handler for the object's logical cluster and invokes it.
+	// c.resourceQuotaInformer.Informer().AddEventHandlerWithResyncPeriod(
+	// 	cache.ResourceEventHandlerFuncs{
+	// 		AddFunc: func(obj interface{}) {
+	// 			h := c.getResourceQuotaEventHandler(obj)
+	// 			if h == nil {
+	// 				return
+	// 			}
+	// 			h.OnAdd(obj)
+	// 		},
+	// 		UpdateFunc: func(oldObj, newObj interface{}) {
+	// 			h := c.getResourceQuotaEventHandler(oldObj)
+	// 			if h == nil {
+	// 				return
+	// 			}
+	// 			h.OnUpdate(oldObj, newObj)
+	// 		},
+	// 		DeleteFunc: func(obj interface{}) {
+	// 			h := c.getResourceQuotaEventHandler(obj)
+	// 			if h == nil {
+	// 				return
+	// 			}
+	// 			h.OnDelete(obj)
+	// 		},
+	// 	},
+	// 	quotaRecalculationPeriod,
+	// )
```

Comment:
debugging code?

Reply:
No, was stuff I didn't want to lose in case some refactoring didn't work. Will be removing.